### PR TITLE
New mmrdiversify

### PR DESCRIPTION
**Pull request by Liusy12138**
Implement `MaximalMarginalRelevanceDiversifier` which using Maximal Marginal Relevance algorithm to diversify ranking result，please refer to issue #53 .

### CODE REVIEW

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/abtesting/experiment/diversify/MMRDiversifyExperiment.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import com.dmetasoul.metaspore.pipeline.annotation.ExperimentAnnotation;
+import com.dmetasoul.metaspore.pipeline.impl.Context;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.Map;
+
+@ExperimentAnnotation(name = "diversify.MMR")
+@Component
+
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/abtesting/experiment/diversify/MMRDiversifyExperiment.java:32]: delete this blank line

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+
+// References:
+// * https://dl.acm.org/doi/10.1145/290941.291025
+
+@Service
+public class MaximalMarginalRelevanceDiversifier implements Diversifier {
+    public static final String DIVERSIFIER_NAME = "MaximalMarginalRelevanceDiversifier";
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java:30]: ALGO_NAME or DIVERSIFIER_NAME?

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    public static final double LAMBDA = 0.7;
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java:31]: please rename to default_lambda or something else.

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                if (genreInWindow.get(deleteGenre) == 0) {
+                    genreInWindow.remove(deleteGenre);
+                }
+            }
+        }
+
+        return itemModels;
+    }
+
+    // simScore={A\cup B}/{A\cup B}
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java:127]: please change this to simScore=\frac{A \cup B}/{A \cup B}

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/SimpleDiversifier.java [hunk 5]
```diff
@@ -16,74 +16,98 @@
...
 import com.dmetasoul.metaspore.demo.movielens.model.ItemModel;
+import com.dmetasoul.metaspore.demo.movielens.model.RecommendContext;
 import org.springframework.stereotype.Service;
-import org.springframework.util.CollectionUtils;
 
 import java.util.*;
 
 @Service
 public class SimpleDiversifier implements Diversifier {
+    public static final String DIVERSIFIER_NAME="SimpleDiersifier";
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/SimpleDiversifier.java:29]: ALGO_NAME or DIVERSIFIER_NAME?

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/SimpleDiversifier.java [hunk 6]
```diff
@@ -16,74 +16,98 @@
...
+            stepCount++;
         }
-        return itemModels;
+        return diverseResult;
     }
 
-    public static HashSet<String> groupByType(List<ItemModel> numbers) {
-        HashSet<String> hashset = new HashSet<>();
-        Map<String, List<ItemModel>> map = new HashMap<>();
+    public static Map<String, Integer> groupByType(List<ItemModel> numbers) {
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/SimpleDiversifier.java:102]: please do not write this function twice, move this into some utils class or parent class.

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/impl/DiversifyServiceImpl.java [hunk 7]
```diff
@@ -18,29 +18,31 @@
...
 import com.dmetasoul.metaspore.demo.movielens.model.ItemModel;
+import com.dmetasoul.metaspore.demo.movielens.model.RecommendContext;
 import org.springframework.stereotype.Service;
 
 import java.util.*;
 
 @Service
 public class DiversifyServiceImpl implements DiversifierService {
-    public static final List<String> DIVERSIFIER_NAMES = List.of("SimpleDiversifier");
+    public static final String DIVERSIFIER_NAMES = "SimpleDiversifier";
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/impl/DiversifyServiceImpl.java:29]: Diversifier_NAME(S)? or something else name?

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/DiverseProvider.java [hunk 8]
```diff
@@ -31,18 +31,10 @@ public DiverseProvider(List<Diversifier> diversifiers) {
...
+    public Diversifier getDiversifiers(String name) {
+        if (name == null) {
+            return diversifierMap.get("simplediersifier");
         }
-        ArrayList<Diversifier> diversifiers = new ArrayList<>();
-        for (String name : names) {
-            diversifiers.add(getDiversifier(name));
-        }
-        return diversifiers;
+        return diversifierMap.get(name.toLowerCase());
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/DiverseProvider.java:38]: why delete 
`public Collection<Diversifier> getDiversifiers(Collection<String> names)`
this function?

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+import com.dmetasoul.metaspore.demo.movielens.diversify.diversifier.Diversifier;
+import com.dmetasoul.metaspore.demo.movielens.model.ItemModel;
+import com.dmetasoul.metaspore.demo.movielens.model.RecommendContext;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+
+// References:
+// * https://dl.acm.org/doi/10.1145/290941.291025
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java:26]: Give the paper name.

**Review by intelligencegear — CHANGES_REQUESTED** [review]

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/model/ItemModel.java [hunk 10]
```diff
@@ -43,6 +48,14 @@ public ItemModel() {
         originalRankingScoreMap = new HashMap<>();
     }
 
+    public void setGenreList(String genre) {
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/model/ItemModel.java:51]: if (StringUtils.isEmpty(genre)) {return ...}

**Review by intelligencegear — COMMENTED** [review]
please add this check.

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/model/ItemModel.java [hunk 11]
```diff
@@ -18,14 +18,17 @@
 
 import com.dmetasoul.metaspore.demo.movielens.domain.Item;
 import com.dmetasoul.metaspore.demo.movielens.domain.ItemFeature;
+import org.codehaus.plexus.util.StringUtils;
 
-import java.util.HashMap;
-import java.util.Optional;
+import java.util.*;
+
+import static com.dmetasoul.metaspore.demo.movielens.common.Constants.SEQUENCE_FEATURE_SPLITTER;
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/model/ItemModel.java:25]: remove this static import.

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/model/ItemModel.java [hunk 12]
```diff
@@ -65,6 +80,7 @@ public String getGenre() {
 
     public void setGenre(String genre) {
         this.genre = genre;
+
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/model/ItemModel.java:83]: remove this blank line.

**Review by intelligencegear — CHANGES_REQUESTED** [review]

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/model/ItemModel.java [hunk 13]
```diff
@@ -16,16 +16,18 @@
 
 package com.dmetasoul.metaspore.demo.movielens.model;
 
+import com.dmetasoul.metaspore.demo.movielens.common.Constants;
 import com.dmetasoul.metaspore.demo.movielens.domain.Item;
 import com.dmetasoul.metaspore.demo.movielens.domain.ItemFeature;
+import org.codehaus.plexus.util.StringUtils;
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/model/ItemModel.java:22]: change this import to google guava or spring libraries.

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/retrieval/impl/RetrievalServiceImpl.java [hunk 14]
```diff
@@ -52,7 +52,7 @@ public List<ItemModel> match(RecommendContext recommendContext, UserModel userMo
         }
 
         List<ItemModel> itemModels = new ArrayList<>();
-        
+
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/retrieval/impl/RetrievalServiceImpl.java:55]: please revert this change.

**Review by intelligencegear — CHANGES_REQUESTED** [review]

**Review by intelligencegear — APPROVED** [review]
please change description of this squash commit.

**Review by andafterall — APPROVED** [review]

**Review by intelligencegear — APPROVED** [review]

**Comment by Liusy12138** [comment]
Associate `New mmrdiversify` with the `[demo] MaximalMarginalRelevanceDiversifier demo`

**Pull request merged**